[Merged by Bors] - feat(order/ideal, data/set/lattice): when order ideals are a complete lattice

atarnoam opened the pull request:
- Added the `ideal_Inter_nonempty` property, which states that the intersection of all ideals in the lattice is nonempty.
- Proved that when a preorder has the above property and is a `semilattice_sup`, its ideals are a complete lattice
- Added some lemmas about empty intersections in set/lattice, akin to #9033

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

Hey, I'm working towards the prime ideal theorem in distributive lattices. This PR is almost all I need!

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

atarnoam committed:
Added is_prime.mem_or_compl_mem

atarnoam committed:
Added is_prime.mem_compl_of_not_mem

atarnoam committed:
Added sup_inf_inf_compl

atarnoam committed:
Update src/order/prime_ideal.lean

Co-authored-by: Bryan Gin-ge Chen <bryangingechen@gmail.com>

atarnoam committed:
Update src/order/prime_ideal.lean

Co-authored-by: Bryan Gin-ge Chen <bryangingechen@gmail.com>

atarnoam committed:
Update src/order/boolean_algebra.lean

Co-authored-by: Bryan Gin-ge Chen <bryangingechen@gmail.com>

atarnoam committed:
- Added is_proper.top_not_mem
- Added is_proper.not_mem_of_compl_mem
- Added is_proper.not_mem_or_compl_not_mem

atarnoam committed:
- Added is_prime_of_mem_or_compl_mem

atarnoam committed:
- Added is_prime_iff_mem_or_compl_mem
- Changed signature of is_prime_of_mem_or_compl_mem

atarnoam committed:
- Added is_prime.is_maximal

atarnoam committed:
cosmetic changes

atarnoam committed:
linted

atarnoam committed:
wip

atarnoam committed:
- Added sInter_eq_empty_iff
- Added sInter_nonempty_iff

atarnoam committed:
wip

atarnoam committed:
wip

atarnoam committed:
- Added theorems on when bInter, Inter are nonempty
- Defined Inf on ideals, and the class ideal_Inter_nonempty

atarnoam committed:
ideals are a complete lattice

atarnoam committed:
Merged with current master

atarnoam committed:
Merge branch 'master' into ideal-lattices

jcommelin commented on src/data/set/lattice.lean:
```diff
@@ -742,10 +742,28 @@ lemma sUnion_eq_univ_iff {c : set (set α)} :
+by simp [set.eq_empty_iff_forall_not_mem]
+
+-- classical
+@[simp] theorem Inter_nonempty_iff {f : ι → set α} : (⋂ i, f i).nonempty ↔ ∃ x, ∀ i, x ∈ f i :=
```
I'm not sure this should be in the global simp-set. I would rather leave it up to the user to step across of the `nonempty` API "boundary".
```suggestion
theorem Inter_nonempty_iff {f : ι → set α} : (⋂ i, f i).nonempty ↔ ∃ x, ∀ i, x ∈ f i :=
```

atarnoam replied:
Looking at the existing theorems, there are (global) simp theorems `nonempty_xUnion`.
Shouldn't we be consistent with it? And also change the names of the theorems to `nonempty_xInter` too?

atarnoam replied:
I did change the name, but still not sure what to do with simp

jcommelin commented on src/data/set/lattice.lean:
```diff
@@ -742,10 +742,28 @@ lemma sUnion_eq_univ_iff {c : set (set α)} :
+by simp [← ne_empty_iff_nonempty, Inter_eq_empty_iff]
+
+-- classical
+@[simp] theorem bInter_nonempty_iff {f : α → set β} {s : set α} :
```
```suggestion
theorem bInter_nonempty_iff {f : α → set β} {s : set α} :
```

jcommelin commented on src/order/ideal.lean:
```diff
@@ -151,6 +168,22 @@ end⟩
+end
+
+lemma ideal_Inter_nonempty_of_exists_all_mem (h : ∃ a : P, ∀ I : ideal P, a ∈ I) :
+ideal_Inter_nonempty P :=
```
```suggestion
  ideal_Inter_nonempty P :=
```

jcommelin commented on src/order/ideal.lean:
```diff
@@ -151,6 +168,22 @@ end⟩
+{ Inter_nonempty := by rwa set.Inter_nonempty_iff }
+
+lemma ideal_Inter_nonempty_iff :
+ideal_Inter_nonempty P ↔ ∃ a : P, ∀ I : ideal P, a ∈ I :=
```
```suggestion
  ideal_Inter_nonempty P ↔ ∃ a : P, ∀ I : ideal P, a ∈ I :=
```

jcommelin commented on src/order/ideal.lean:
```diff
@@ -299,14 +336,80 @@ end
+    cases ideal_Inter_nonempty.exists_all_mem ‹_› with a ha,
+    exact ⟨a, ha _, ha _⟩
+  end
+}
```
```suggestion
  end }
```

jcommelin commented on src/order/ideal.lean:
```diff
@@ -299,14 +336,80 @@ end
+variables {α β γ : Type*} {ι : Sort*}
+
+lemma ideal_Inter_nonempty.ideal_Inter_nonempty {f : ι → ideal P} :
+(⋂ x, (f x : set P)).nonempty :=
```
Please indent the *type* of statements by 2 spaces. Also elsewhere.

eric-wieser commented on src/order/ideal.lean:
```diff
@@ -85,6 +90,18 @@ lemma inter_nonempty [preorder P] [ideal_inter_nonempty P] :
+    satisfies that its ideal poset is a complete lattice.
+-/
+class ideal_Inter_nonempty (P) [preorder P] : Prop :=
+(Inter_nonempty : (⋂ (I : ideal P), I.carrier).nonempty)
```
```suggestion
(Inter_nonempty : (⋂ (I : ideal P), (I : set P)).nonempty)
```

atarnoam replied:
This actually doesn't work because `has_coe` is only defined later (in the preorder section).
Should I keep this with `I.carrier`, move has_coe earlier or move the definition to the preorder section?

kbuzzard replied:
Another possibility (probably worse) is to call the field `Inter_nonempty'` and then define `ideal_Inter_nonempty.Inter_nonempty` later. But sure, if you can move it to somewhere where it does work then this seems like the nicest solution.

kbuzzard replied:
Looking more carefully, why not just move `ideal_Inter_nonempty` and `Inter_nonempty` to just before you start to use them on line 170 or so?

eric-wieser commented on src/order/ideal.lean:
```diff
@@ -299,14 +336,80 @@ end
+    end }
+
+instance : has_Inf (ideal P) :=
+{ Inf := ideal.Inf }
```
I think you should inline the implementation directly into the `has_Inf`. If you're looking to make this shorter, perhaps extracted `directed` as a standalone lemma.

atarnoam committed:
Apply suggestions from code review

atarnoam committed:
Apply suggestions from code review

atarnoam committed:
Apply suggestions from code review

atarnoam committed:
Inline has_Inf

eric-wieser commented on src/order/ideal.lean:
```diff
@@ -151,6 +168,22 @@ end⟩
 @[mk_iff] class is_maximal (I : ideal P) extends is_proper I : Prop :=
 (maximal_proper : ∀ ⦃J : ideal P⦄, I < J → J.carrier = set.univ)
 
+lemma ideal_Inter_nonempty.exists_all_mem (hP : ideal_Inter_nonempty P) :
```
Shouldn't there be a version of this expressed with a typeclass argument?
```lean
lemma exists_all_mem [ideal_Inter_nonempty P] :
```

atarnoam replied:
I tried this but this doesn't work easily with lemmas that use this form (for example, changing the proof of `ideal_Inter_nonempty.ideal_bInter_nonempty` to use this just doesn't work for some reason), so I think it is better to not include it.

kbuzzard replied:
It's possible to make Eric's suggestion work. 

```
lemma ideal_Inter_nonempty_iff :
  ideal_Inter_nonempty P ↔ ∃ a : P, ∀ I : ideal P, a ∈ I :=
⟨λ h, by exactI ideal_Inter_nonempty.exists_all_mem, ideal_Inter_nonempty_of_exists_all_mem⟩
```

and then later on (three times) use `obtain` e.g.

```
instance ideal_Inter_nonempty.ideal_inter_nonempty : ideal_inter_nonempty P :=
{ inter_nonempty := λ _ _, begin
    obtain ⟨a, ha⟩ : ∃ a : P, _ := ideal_Inter_nonempty.exists_all_mem,
    exact ⟨a, ha _, ha _⟩,
  end }
```

eric-wieser commented on src/order/ideal.lean:
```diff
@@ -299,14 +336,76 @@ end
+variables {α β γ : Type*} {ι : Sort*}
+
+lemma ideal_Inter_nonempty.ideal_Inter_nonempty {f : ι → ideal P} :
+  (⋂ x, (f x : set P)).nonempty :=
```
The naming here is odd and doesn't indicate how this differs from `Inter_nonempty` above.

atarnoam replied:
This is analogous to `ideal_bInter_nonempty` and is a generalization of `Inter_nonempty`. Should I just remove it? How should it be renamed?

atarnoam committed:
Now mathlib compiles

atarnoam committed:
- Changed ```xInter_nonempty_iff``` to ```nonempty_xInter``` (to comply with ```nonempty_xUnion```)
- Changed ```ideal_Inter_nonempty.ideal_xInter_nonempty``` to ```ideal_Inter_nonempty.all_Inter_nonempty``` (emphasizing the generalization)

atarnoam committed:
- Moved ideal_inter_noenmpty and ideal_Inter_nonempty down
- Changed lemmas about top to use ```coe``` instead of carrier
- ideal_Inter_nonempty.exists_all_mem now uses a typeclass argument

atarnoam committed:
Linted, removed top_eq_univ

YaelDillies commented on src/order/ideal.lean:
```diff
@@ -6,6 +6,9 @@ Authors: David Wärn
 import order.atoms
+import order.bounded_lattice
+import order.complete_lattice
+import order.zorn
```
You don't need to add any imports.
```suggestion
import data.equiv.encodable.basic
import order.atoms
```

YaelDillies commented on src/order/ideal.lean:
```diff
@@ -299,14 +339,76 @@ end
+@[simp] lemma mem_Inf : x ∈ Inf s ↔ ∀ I ∈ s, x ∈ I :=
+by { change x ∈ (⋂ (I ∈ s), (I : set P)) ↔ ∀ I ∈ s, x ∈ I, simp }
+
+@[simp] lemma Inf_coe : ↑(Inf s) = ⋂ (I ∈ s), (I : set P) := rfl
```
```suggestion
@[simp] lemma coe_Inf : ↑(Inf s) = ⋂ (I ∈ s), (I : set P) := rfl
```
`Inf_coe` would be about `Inf ↑s`.

YaelDillies commented on src/order/ideal.lean:
```diff
@@ -178,17 +220,15 @@ instance : order_top (ideal P) :=
   .. ideal.partial_order }
 
-@[simp] lemma top_carrier : (⊤ : ideal P).carrier = set.univ :=
+@[simp] lemma top_coe : ((⊤ : ideal P) : set P) = set.univ :=
```
```suggestion
@[simp] lemma coe_top : ((⊤ : ideal P) : set P) = set.univ :=
```

YaelDillies reviewed (commented):
Is there no use to the dual theory of filters?

atarnoam committed:
Changed top_coe to coe_top

atarnoam committed:
Apply suggestions from code review

Co-authored-by: Yaël Dillies <yael.dillies@gmail.com>

bryangingechen commented on src/order/ideal.lean:
```diff
@@ -149,7 +138,52 @@ end⟩
+  ∀ (I J : ideal P), ((I : set P) ∩ (J : set P)).nonempty :=
+ideal_inter_nonempty.inter_nonempty
+
+
```
```suggestion
```

bryangingechen commented on src/order/ideal.lean:
```diff
@@ -299,14 +335,76 @@ end
+      simp [← hS],
+      assume hI,
+      rw set.mem_bInter_iff at *,
+      tauto,
```
```suggestion
      simp only [←hS, sup_mem_iff, mem_coe, set.mem_Inter],
      intro hI,
      rw set.mem_bInter_iff at *,
      exact ⟨hx _ hI, hy _ hI⟩
```

bryangingechen commented on src/order/ideal.lean:
```diff
@@ -299,14 +335,76 @@ end
+lemma le_Inf (h : ∀ J ∈ s, I ≤ J) : I ≤ Inf s :=
+λ _ _, by { simp, tauto }
+
+lemma is_glb_Inf : is_glb s (Inf s) := ⟨assume a, Inf_le, assume a, le_Inf⟩
```
```suggestion
λ _ _, by { simp only [mem_coe, coe_Inf, set.mem_Inter], tauto }

lemma is_glb_Inf : is_glb s (Inf s) := ⟨λ _, Inf_le, λ _, le_Inf⟩
```

bryangingechen reviewed (commented):
These suggestions just remove some non-terminal `simp`s and change some `assume`s for consistency.

Any other final comments? @kbuzzard @jcommelin @eric-wieser

atarnoam committed:
Apply suggestions from code review

Co-authored-by: Bryan Gin-ge Chen <bryangingechen@gmail.com>

bryangingechen reviewed (approved):
Thanks!
bors r+
